src/extension/rpc.ts

```diff
@@ -0,0 +1,191 @@
+import { NoInfer, SafeAny } from "../types";
+import { MessageAdapter } from "./messageAdapters";
+import {
+  ApolloClientDevtoolsRPCMessage,
+  MessageType,
+  isRPCMessage,
+} from "./messages";
+
+type RPCParams = Record<string, unknown>;
+
+type RPCRequestMessageFormat = {
+  type: string;
+  params: RPCParams;
+};
+
+type RPCResponseMessageFormat =
+  | { sourceId: number; result: unknown }
+  | { sourceId: number; error: unknown };
+
+export type RPCMessage<
+  Name extends string,
+  Params extends RPCParams,
+  ReturnType,
+> = {
+  __name: Name;
+  __params: Params;
+  __returnType: ReturnType;
+};
+
+export interface RpcClient<
+  Messages extends RPCMessage<string, RPCParams, SafeAny>,
+> {
+  request: <TName extends Messages["__name"]>(
+    name: TName,
+    params: Extract<Messages, { __name: TName }>["__params"],
+    options?: { timeoutMs?: number }
+  ) => Promise<Extract<Messages, { __name: TName }>["__returnType"]>;
+}
+
+let nextMessageId = 0;
+const DEFAULT_TIMEOUT = 30_000;
+
+export function createRpcClient<
+  Messages extends RPCMessage<string, RPCParams, SafeAny>,
+>(
+  adapter: MessageAdapter<
+    ApolloClientDevtoolsRPCMessage<RPCRequestMessageFormat>
+  >
+): RpcClient<Messages> {
+  return {
+    request: (name, params, options) => {
+      return Promise.race([
+        new Promise((_, reject) => {
+          setTimeout(
+            () => reject(new Error("Timeout waiting for message")),
+            options?.timeoutMs ?? DEFAULT_TIMEOUT
+          );
+        }),
+        new Promise((resolve, reject) => {
+          const id = ++nextMessageId;
+
+          const removeListener = adapter.addListener((message) => {
+            if (
+              isRPCMessage(message) &&
+              "sourceId" in message.payload &&
+              message.payload.sourceId === id
+            ) {
+              if ("error" in message.payload) {
+                reject(message.payload.error);
+              } else {
+                resolve(message.payload.result);
+              }
+
+              removeListener();
+            }
+          });
+
+          adapter.postMessage({
+            source: "apollo-client-devtools",
+            type: MessageType.RPC,
+            id,
+            payload: { type: name, params },
+          });
+        }),
+      ]);
+    },
+  };
+}
```

Comment:
I would suggest that we try to avoid `Promise.race` here as it is very prone to [memory leaks](https://github.com/nodejs/node/issues/17469#issuecomment-685216777).


```suggestion
export function createRpcClient<
  Messages extends RPCMessage<string, RPCParams, SafeAny>,
>(
  adapter: MessageAdapter<
    ApolloClientDevtoolsRPCMessage<RPCRequestMessageFormat>
  >
): RpcClient<Messages> {
  return {
    request: (name, params, options) => {
      return new Promise((resolve, reject) => {
        const id = ++nextMessageId;

        const timeout = setTimeout(() => {
          removeListener();
          reject(new Error("Timeout waiting for message"));
        }, options?.timeoutMs ?? DEFAULT_TIMEOUT);

        const removeListener = adapter.addListener((message) => {
          if (
            isRPCMessage(message) &&
            "sourceId" in message.payload &&
            message.payload.sourceId === id
          ) {
            if ("error" in message.payload) {
              reject(message.payload.error);
            } else {
              resolve(message.payload.result);
            }

            removeListener();
            clearTimeout(timeout);
          }
        });

        adapter.postMessage({
          source: "apollo-client-devtools",
          type: MessageType.RPC,
          id,
          payload: { type: name, params },
        });
      });
    },
  };
}
```

Reply:
Also, writing it like this makes it easier to call `removeListener` in the case of a timeout.

Reply:
Updated in [cb7169f](https://github.com/apollographql/apollo-client-devtools/pull/1256/commits/cb7169fabdbca765650c8882e21d737415228357)